File: app/services/analytics/insights.py
from __future__ import annotations

from dataclasses import dataclass
from app.services.analytics.engine import SpendingAnalytics


@dataclass
class Insight:
    id: str
    type: str
    severity: str  # high | medium | low | info
    title: str
    description: str
    potential_saving: float | None
    category: str | None


class InsightsGenerator:
    def generate(self, analytics: SpendingAnalytics) -> list[Insight]:
        insights: list[Insight] = []
        income = analytics.total_income
        spend = analytics.total_expenditure
        idx = 0

        def make_id() -> str:
            nonlocal idx
            idx += 1
            return str(idx)

        # Savings rate
        if analytics.savings_rate >= 20:
            insights.append(Insight(
                id=make_id(), type="positive", severity="info",
                title=f"Strong savings rate: {analytics.savings_rate:.1f}%",
                description=f"You saved ₦{analytics.net_cash_flow:,.0f} this period. That's a great habit!",
                potential_saving=None, category=None,
            ))
        elif analytics.savings_rate < 5 and income > 0:
            insights.append(Insight(
                id=make_id(), type="low_savings", severity="high",
                title=f"Low savings rate: {analytics.savings_rate:.1f}%",
                description=f"Less than 5% of income was saved this period. Aim for at least 20%.",
                potential_saving=round(income * 0.20 - analytics.net_cash_flow, 2),
                category=None,
            ))

        # Bank charges
        if income > 0 and analytics.bank_charges_total / income > 0.02:
            insights.append(Insight(
                id=make_id(), type="bank_charges", severity="medium",
                title=f"Bank charges: ₦{analytics.bank_charges_total:,.0f}",
                description=f"You paid ₦{analytics.bank_charges_total:,.0f} in fees "
                            f"({analytics.bank_charges_total / income * 100:.1f}% of income).",
                potential_saving=round(analytics.bank_charges_total * 0.5, 2),
                category="Financial",
            ))

        # Category insights
        for cat_stat in analytics.by_category[:5]:
            # High spending category
            if spend > 0 and cat_stat.percentage > 25 and cat_stat.category != "People":
                saving = cat_stat.total * 0.25
                insights.append(Insight(
                    id=make_id(), type="high_category", severity="medium",
                    title=f"{cat_stat.category} is {cat_stat.percentage:.1f}% of spending",
                    description=f"You spent ₦{cat_stat.total:,.0f} on {cat_stat.category}. "
                                f"Reducing by 25% could save ₦{saving:,.0f}/month.",
                    potential_saving=round(saving, 2),
                    category=cat_stat.category,
                ))

        # Subscriptions
        if analytics.subscription_total > 0 and income > 0:
            sub_pct = analytics.subscription_total / income * 100
            if sub_pct > 10:
                saving = analytics.subscription_total * 0.25
                insights.append(Insight(
                    id=make_id(), type="subscriptions", severity="medium",
                    title=f"Subscriptions cost ₦{analytics.subscription_total:,.0f}/month",
                    description=f"That's {sub_pct:.1f}% of income — ₦{analytics.subscription_total * 12:,.0f}/year. "
                                f"Review unused ones.",
                    potential_saving=round(saving, 2),
                    category="Subscriptions",
                ))

        # Largest debit
        if analytics.largest_debit:
            insights.append(Insight(
                id=make_id(), type="largest_transaction", severity="info",
                title=f"Biggest spend: ₦{float(analytics.largest_debit.get('amount', 0)):,.0f}",
                description=f"{analytics.largest_debit.get('description', 'Unknown')[:80]}",
                potential_saving=None, category=analytics.largest_debit.get("category"),
            ))

        return insights


insights_generator = InsightsGenerator()

File: src/services/analytics/insights.ts
import type { GeneratedInsight, NormalizedTransaction, SpendingAnalytics } from "../../types";

// ─── Insights Generator — deterministic, data-driven ──────────────────────────

export class InsightsGenerator {
  generate(
    transactions: NormalizedTransaction[],
    analytics: SpendingAnalytics,
    previous?: NormalizedTransaction[]
  ): GeneratedInsight[] {
    const insights: GeneratedInsight[] = [];

    // ── Biggest spending day
    const biggestDay = analytics.byDay.reduce(
      (best, d) => (d.expenditure > best.expenditure ? d : best),
      { date: "", expenditure: 0, income: 0, count: 0 }
    );
    if (biggestDay.expenditure > 0) {
      insights.push({
        type: "BIGGEST_SPENDING_DAY",
        title: "Biggest spending day",
        body: `${this.formatDate(biggestDay.date)} was your biggest spending day with ₦${this.fmt(biggestDay.expenditure)} across ${biggestDay.count} transaction${biggestDay.count !== 1 ? "s" : ""}.`,
        icon: "🔥",
        severity: "info",
        data: biggestDay,
      });
    }

    // ── Largest transaction
    const allDebits = transactions
      .filter((t) => t.type === "debit")
      .sort((a, b) => b.amount - a.amount);
    if (allDebits.length > 0) {
      const largest = allDebits[0];
      insights.push({
        type: "LARGEST_TRANSACTION",
        title: "Largest transaction",
        body: `Your largest transaction was ₦${this.fmt(largest.amount)} — ${largest.description}.`,
        icon: "💸",
        severity: "info",
        data: { amount: largest.amount, description: largest.description, date: largest.date },
      });
    }

    // ── Top spending category
    if (analytics.byCategory.length > 0) {
      const top = analytics.byCategory[0];
      insights.push({
        type: "TOP_CATEGORY",
        title: `${top.category} is your top category`,
        body: `${top.category} was your biggest spending category at ₦${this.fmt(top.amount)} (${Math.round(top.percentage)}% of total spending).`,
        icon: "📊",
        severity: "info",
        data: top,
      });
    }

    // ── Top recipient (person)
    const people = analytics.byEntity
      .filter((e) => e.totalSent > 0)
      .sort((a, b) => b.totalSent - a.totalSent);
    if (people.length > 0) {
      const topPerson = people[0];
      insights.push({
        type: "TOP_RECIPIENT",
        title: `${topPerson.entity} received the most`,
        body: `You sent ₦${this.fmt(topPerson.totalSent)} to ${topPerson.entity} across ${topPerson.count} transaction${topPerson.count !== 1 ? "s" : ""}.`,
        icon: "👤",
        severity: "info",
        data: topPerson,
      });
    }

    // ── Subscription total
    if (analytics.subscriptionTotal > 0) {
      const annual = analytics.subscriptionTotal * 12;
      insights.push({
        type: "SUBSCRIPTION_TOTAL",
        title: "Your subscriptions this period",
        body: `Your active subscriptions cost ₦${this.fmt(analytics.subscriptionTotal)} this period — that's approximately ₦${this.fmt(annual)} per year.`,
        icon: "🔄",
        severity: analytics.subscriptionTotal > analytics.totalIncome * 0.15 ? "warning" : "info",
        data: {
          monthly: analytics.subscriptionTotal,
          annual,
          percentage: analytics.totalIncome > 0
            ? Math.round((analytics.subscriptionTotal / analytics.totalIncome) * 100)
            : 0,
        },
      });
    }

    // ── Month-over-month comparison
    if (previous && previous.length > 0) {
      const prevSpend = previous
        .filter((t) => t.type === "debit")
        .reduce((s, t) => s + t.amount, 0);
      const currSpend = analytics.totalExpenditure;

      if (prevSpend > 0) {
        const change = ((currSpend - prevSpend) / prevSpend) * 100;
        const direction = change > 0 ? "increased" : "decreased";
        const icon = change > 0 ? "📈" : "📉";
        const severity = change > 20 ? "warning" : change < -10 ? "success" : "info";

        insights.push({
          type: "MOM_CHANGE",
          title: `Spending ${direction} vs last period`,
          body: `Your spending ${direction} by ${Math.abs(Math.round(change))}% compared with the previous period (₦${this.fmt(prevSpend)} → ₦${this.fmt(currSpend)}).`,
          icon,
          severity,
          data: { prevSpend, currSpend, changePercent: Math.round(change * 10) / 10 },
        });
      }
    }

    // ── Savings rate
    if (analytics.savingsRate > 0) {
      const severity =
        analytics.savingsRate >= 20 ? "success" : analytics.savingsRate >= 10 ? "info" : "warning";
      insights.push({
        type: "SAVINGS_RATE",
        title: "Your savings rate",
        body: `You saved ${analytics.savingsRate.toFixed(1)}% of your income this period (₦${this.fmt(analytics.netCashFlow)} net cash flow).`,
        icon: analytics.savingsRate >= 20 ? "💰" : "💡",
        severity,
        data: {
          savingsRate: analytics.savingsRate,
          netCashFlow: analytics.netCashFlow,
          totalIncome: analytics.totalIncome,
        },
      });
    }

    // ── Bank charges
    if (analytics.bankChargesTotal > 0) {
      insights.push({
        type: "BANK_CHARGES",
        title: "Banking fees this period",
        body: `You paid ₦${this.fmt(analytics.bankChargesTotal)} in banking fees and charges. Over a year, that's approximately ₦${this.fmt(analytics.bankChargesTotal * 12)}.`,
        icon: "🏦",
        severity: analytics.bankChargesTotal > analytics.totalIncome * 0.02 ? "warning" : "info",
        data: { bankCharges: analytics.bankChargesTotal, annualEstimate: analytics.bankChargesTotal * 12 },
      });
    }

    return insights;
  }

  private fmt(amount: number): string {
    return new Intl.NumberFormat("en-NG").format(Math.round(amount));
  }

  private formatDate(dateStr: string): string {
    if (!dateStr) return "Unknown date";
    const d = new Date(dateStr);
    return d.toLocaleDateString("en-NG", { weekday: "long", day: "numeric", month: "long" });
  }
}

export const insightsGenerator = new InsightsGenerator();
